intake_xarray/image.py

```diff
@@ -0,0 +1,147 @@
+import xarray as xr
+from .base import XarraySource
+
+
+def add_leading_dimension(x):
+    return x[None, ...]
+
+
+def dask_imread(filename, imread=None, preprocess=None):
+    """ Read a stack of images into a dask array
+    Parameters
+    ----------
+    filename: string, iter
+        A globstring like 'myfile.*.png', a list of files, or anything
+        that can go into imread.
+    imread: function (optional)
+        Optionally provide custom imread function.
+        Function should expect a filename and produce a numpy array.
+        Defaults to ``skimage.io.imread``.
+    preprocess: function (optional)
+        Optionally provide custom function to preprocess the image.
+        Function should expect a numpy array for a single image.
+    Examples
+    --------
+    >>> from dask.array.image import imread
+    >>> im = imread('2015-*-*.png')  # doctest: +SKIP
+    >>> im.shape  # doctest: +SKIP
+    (365, 1000, 1000, 3)
+    Returns
+    -------
+    Dask array of all images stacked along the first dimension.  All images
+    will be treated as individual chunks
+    """
+    import os
+    from glob import glob
+
+    try:
+        from skimage.io import imread as sk_imread
+    except (AttributeError, ImportError):
+        pass
+
+    from dask.array import Array
+    from dask.base import tokenize
+
+    imread = imread or sk_imread
+    name = None
+
+    if isinstance(filename, list):
+        filenames = filename
+    elif '*' in filename:
+        filenames = sorted(glob(filename))
```

Comment:
ks_imread definitely can read from arbitrary file-like objects, so this case should certainly use open_files, so we can directly access s3, gcs... . The only wrinkle is, that the files are not really opened until coming to a `with` block, so you'll need a little helper function like

```
def read_open_file(open_file, imread):
    with open_file as f:
        return imread(f)
```
and we wish to specify, if the user is to provide their own function, that it must either take any file-like